File: frontend/src/lib/validators/expense.ts
import { z } from "zod";

export const expenseSchema = z.object({
  title: z
    .string()
    .min(1, "Title is required")
    .max(50, "Title must be 50 characters or less")
    .refine((val) => val.trim().length > 0, "Title cannot be empty or whitespace only"),
  category_id: z
    .string()
    .uuid("Please select a valid category"),
  amount: z
    .union([z.number(), z.string()])
    .transform((val) => {
      if (typeof val === "number") return val;
      const parsed = parseFloat(val);
      return isNaN(parsed) ? 0 : parsed;
    })
    .refine((val) => val > 0, "Amount must be a positive number"),
  expense_date: z
    .string()
    .min(1, "Date is required")
    .refine((val) => {
      const selectedDate = new Date(val);
      // Remove time components for strict date comparison
      const today = new Date();
      today.setHours(23, 59, 59, 999); // allow today's date regardless of time zones
      return selectedDate <= today;
    }, "Date cannot be in the future"),
  notes: z
    .string()
    .max(250, "Notes must be 250 characters or less")
    .optional()
    .nullable()
    .or(z.literal("")),
  payment_mode: z
    .enum(["cash", "card", "upi", "other"])
    .optional()
    .nullable()
    .or(z.literal("")),
});

export type ExpenseFormValues = z.infer<typeof expenseSchema>;

File: backend/app/schemas/expense.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Literal
from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.category import CategoryBriefResponse


class ExpenseBase(BaseModel):
    title: str = Field(..., max_length=50)
    category_id: uuid.UUID
    amount: Decimal = Field(..., gt=0)
    expense_date: date
    notes: str | None = Field(None, max_length=250)
    payment_mode: Literal["cash", "card", "upi", "other"] | None = None

    @field_validator("title")
    @classmethod
    def validate_title(cls, v: str) -> str:
        stripped = v.strip()
        if not stripped:
            raise ValueError("Title cannot be empty or whitespace only")
        return stripped

    @field_validator("expense_date")
    @classmethod
    def validate_date(cls, v: date) -> date:
        if v > date.today():
            raise ValueError("Date cannot be in the future")
        return v


class ExpenseCreate(ExpenseBase):
    pass


class ExpenseUpdate(BaseModel):
    title: str | None = Field(None, max_length=50)
    category_id: uuid.UUID | None = None
    amount: Decimal | None = Field(None, gt=0)
    expense_date: date | None = None
    notes: str | None = Field(None, max_length=250)
    payment_mode: Literal["cash", "card", "upi", "other"] | None = None

    @field_validator("title")
    @classmethod
    def validate_title(cls, v: str | None) -> str | None:
        if v is None:
            return v
        stripped = v.strip()
        if not stripped:
            raise ValueError("Title cannot be empty or whitespace only")
        return stripped

    @field_validator("expense_date")
    @classmethod
    def validate_date(cls, v: date | None) -> date | None:
        if v is None:
            return v
        if v > date.today():
            raise ValueError("Date cannot be in the future")
        return v


class ExpenseResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    title: str
    amount: Decimal
    expense_date: date
    notes: str | None
    payment_mode: str | None
    category_id: uuid.UUID
    category: CategoryBriefResponse
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)
